File: toggly-ruby/toggly/lib/toggly/http_request_mapper.rb
# frozen_string_literal: true

module Toggly
  # Maps common HTTP headers into RequestContext fields.
  #
  # Does not invent identity, groups, or claims — merge those separately.
  class HttpRequestMapper
    COUNTRY_HEADERS = %w[
      cf-ipcountry
      x-vercel-ip-country
      cloudfront-viewer-country
    ].freeze

    class << self
      # Build RequestContext from a header bag (case-insensitive keys).
      #
      # @param headers [Hash, nil]
      # @return [RequestContext]
      def from_http_headers(headers)
        return RequestContext.new if headers.nil? || headers.empty?

        RequestContext.new(
          user_agent: header(headers, "user-agent"),
          accept_language: header(headers, "accept-language"),
          country: first_present(headers, *COUNTRY_HEADERS)
        )
      end

      # Merge HTTP-mapped request fields over an existing evaluation context.
      #
      # @param headers [Hash, nil]
      # @param base [Context, nil]
      # @return [Context]
      def merge_into(headers, base)
        request = from_http_headers(headers)
        return Context.new(request: request) if base.nil?

        base.with_request(request)
      end

      private

      def first_present(headers, *names)
        names.each do |name|
          value = header(headers, name)
          return value if value
        end
        nil
      end

      def header(headers, name)
        lower = name.downcase
        headers.each do |key, value|
          next unless key && key.to_s.downcase == lower
          next if value.nil?

          text = value.to_s
          return text unless text.empty?
        end
        nil
      end
    end
  end
end

File: Toggly.FeatureManagement.Java/toggly-core/src/main/java/io/toggly/core/context/HttpRequestMapper.java
package io.toggly.core.context;

import java.util.Locale;
import java.util.Map;

/**
 * Maps common HTTP request headers into {@link RequestContext} fields.
 *
 * <p>Does not invent identity, groups, or claims — merge those separately.</p>
 */
public final class HttpRequestMapper {

    private HttpRequestMapper() {}

    /**
     * Builds a {@link RequestContext} from a header bag (case-insensitive keys).
     *
     * @param headers header name → value (first value wins for multi-valued maps)
     * @return request context with mapped fields (any may be null)
     */
    public static RequestContext fromHttpHeaders(Map<String, String> headers) {
        if (headers == null || headers.isEmpty()) {
            return RequestContext.of(null, null, null);
        }
        return RequestContext.builder()
                .userAgent(header(headers, "user-agent"))
                .acceptLanguage(header(headers, "accept-language"))
                .country(firstPresent(headers,
                        "cf-ipcountry",
                        "x-vercel-ip-country",
                        "cloudfront-viewer-country"))
                .build();
    }

    /**
     * Merges HTTP-mapped request fields over an existing evaluation context.
     *
     * @param headers HTTP headers
     * @param base existing context (identity/groups/claims/traits/entity preserved)
     * @return a new context with {@code request} set from headers
     */
    public static EvaluationContext mergeInto(Map<String, String> headers, EvaluationContext base) {
        RequestContext request = fromHttpHeaders(headers);
        if (base == null) {
            return EvaluationContext.builder().request(request).build();
        }
        return base.withRequest(request);
    }

    private static String firstPresent(Map<String, String> headers, String... names) {
        for (String name : names) {
            String value = header(headers, name);
            if (value != null) {
                return value;
            }
        }
        return null;
    }

    private static String header(Map<String, String> headers, String name) {
        String lower = name.toLowerCase(Locale.ROOT);
        for (Map.Entry<String, String> entry : headers.entrySet()) {
            if (entry.getKey() != null
                    && entry.getKey().toLowerCase(Locale.ROOT).equals(lower)) {
                String value = entry.getValue();
                if (value != null && !value.isEmpty()) {
                    return value;
                }
            }
        }
        return null;
    }
}
